Firestore/Source/API/FSTUserDataReader.mm

```diff
@@ -277,51 +318,70 @@ - (FieldValue)parsedQueryValue:(id)input allowArrays:(bool)allowArrays {
 }
 
-- (FieldValue)parseDictionary:(NSDictionary<NSString *, id> *)dict
-                      context:(ParseContext &&)context {
+- (google_firestore_v1_Value)parseDictionary:(NSDictionary<NSString *, id> *)dict
+                                     context:(ParseContext &&)context {
+  google_firestore_v1_Value result;
+  result.which_value_type = google_firestore_v1_Value_map_value_tag;
+
+  __block std::vector<google_firestore_v1_MapValue_FieldsEntry> fields;
+
   if (dict.count == 0) {
     const FieldPath *path = context.path();
     if (path && !path->empty()) {
       context.AddToFieldMask(*path);
     }
-    return ObjectValue::Empty().AsFieldValue();
   } else {
-    __block ObjectValue result = ObjectValue::Empty();
-
+    // Populate a vector of fields since we don't know the size of the final fields array.
     [dict enumerateKeysAndObjectsUsingBlock:^(NSString *key, id value, BOOL *) {
-      absl::optional<FieldValue> parsedValue =
+      absl::optional<google_firestore_v1_Value> parsedValue =
           [self parseData:value context:context.ChildContext(util::MakeString(key))];
       if (parsedValue) {
-        FieldPath path = FieldPath{util::MakeString(key)};
-        result = result.Set(path, *parsedValue);
+        google_firestore_v1_MapValue_FieldsEntry fieldsEntry;
+        fieldsEntry.key = nanopb::MakeBytesArray(util::MakeString(key));
+        fieldsEntry.value = *parsedValue;
+        fields.push_back(fieldsEntry);
       }
     }];
+  }
 
-    return result;
+  result.map_value.fields_count = static_cast<pb_size_t>(fields.size());
+  result.map_value.fields =
+      nanopb::MakeArray<google_firestore_v1_MapValue_FieldsEntry>(result.map_value.fields_count);
+
+  for (size_t i = 0; i < fields.size(); ++i) {
+    result.map_value.fields[i] = fields[i];
   }
+
+  return result;
 }
 
-- (FieldValue)parseArray:(NSArray<id> *)array context:(ParseContext &&)context {
-  __block FieldValue::Array result;
-  result.reserve(array.count);
+- (google_firestore_v1_Value)parseArray:(NSArray<id> *)array context:(ParseContext &&)context {
+  __block google_firestore_v1_Value result;
+  result.which_value_type = google_firestore_v1_Value_array_value_tag;
+  result.array_value.values_count = static_cast<pb_size_t>([array count]);
+  result.array_value.values =
+      nanopb::MakeArray<google_firestore_v1_Value>(result.array_value.values_count);
 
   [array enumerateObjectsUsingBlock:^(id entry, NSUInteger idx, BOOL *) {
-    absl::optional<FieldValue> parsedEntry = [self parseData:entry
-                                                     context:context.ChildContext(idx)];
+    absl::optional<google_firestore_v1_Value> parsedEntry =
+        [self parseData:entry context:context.ChildContext(idx)];
     if (!parsedEntry) {
       // Just include nulls in the array for fields being replaced with a sentinel.
-      parsedEntry = FieldValue::Null();
+      parsedEntry = absl::optional<google_firestore_v1_Value>{[self encodeNullValue]};
```

Comment:
Nit: explicitly specifying `absl::optional` shouldn't be necessary.

Reply:
Removed